homeassistant/components/google_drive/api.py

```diff
@@ -0,0 +1,206 @@
+        """Create Home Assistant folder if it doesn't exist."""
+        fields = "id,name"
+        query = " and ".join(
+            [
+                "properties has { key='ha' and value='root' }",
+                f"properties has {{ key='instance_id' and value='{self._ha_instance_id}' }}",
+                "trashed=false",
+            ]
+        )
+        res = await self._api.list_files(
+            params={"q": query, "fields": f"files({fields})"}
+        )
+        for file in res["files"]:
+            _LOGGER.debug("Found existing folder: %s", file)
+            return str(file["id"]), str(file["name"])
+
+        file_metadata = {
+            "name": "Home Assistant",
+            "mimeType": "application/vnd.google-apps.folder",
+            "properties": {
+                "ha": "root",
+                "instance_id": self._ha_instance_id,
+            },
+        }
+        _LOGGER.debug("Creating new folder with metadata: %s", file_metadata)
+        res = await self._api.create_file(params={"fields": fields}, json=file_metadata)
+        _LOGGER.debug("Created folder: %s", res)
+        return str(res["id"]), str(res["name"])
+
+    async def async_upload_backup(
+        self,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+    ) -> None:
+        """Upload a backup."""
+        folder_id, _ = await self.async_create_ha_root_folder_if_not_exists()
+        backup_metadata = {
+            "name": f"{backup.name} {backup.date}.tar",
+            "description": json.dumps(backup.as_dict()),
+            "parents": [folder_id],
+            "properties": {
+                "ha": "backup",
+                "instance_id": self._ha_instance_id,
+                "backup_id": backup.backup_id,
+            },
+        }
+        _LOGGER.debug(
+            "Uploading backup: %s with Google Drive metadata: %s",
+            backup.backup_id,
+            backup_metadata,
+        )
+        await self._api.upload_file(
+            backup_metadata,
+            open_stream,
+            timeout=ClientTimeout(total=_UPLOAD_AND_DOWNLOAD_TIMEOUT),
+        )
+        _LOGGER.debug(
+            "Uploaded backup: %s to: '%s'",
+            backup.backup_id,
+            backup_metadata["name"],
+        )
+
+    async def async_list_backups(self) -> list[AgentBackup]:
+        """List backups."""
+        query = " and ".join(
+            [
+                "properties has { key='ha' and value='backup' }",
+                f"properties has {{ key='instance_id' and value='{self._ha_instance_id}' }}",
+                "trashed=false",
+            ]
+        )
+        res = await self._api.list_files(
+            params={"q": query, "fields": "files(description)"}
+        )
+        backups = []
+        for file in res["files"]:
+            backup = AgentBackup.from_dict(json.loads(file["description"]))
```

Comment:
We need to make the AgentBackup serializer / deserializer versioned (not a comment for this PR, it's an improvement of the core backup integration, for a separate PR)